### PR TITLE
Use docker run instead of extra docker build

### PR DESCRIPTION
**Pull request by Xuanwo**
This PR is a try to remove the docker build steps.

### CODE REVIEW

**Comment by pengx17** [comment]
I am not sure if the steps in GitHub Action can be cached? Logseq build itself will take around 12 minutes, which means if use native workflow a single publish will take at least 12 minutes, compared to ~2 min with the dockerized approach.

**Comment by Xuanwo** [comment]
> I am not sure if the steps in GitHub Action can be cached? 

Yes.

GHA supports cache some specified paths, is it enough? Maybe we can cache the build result of logseq?

**Comment by pengx17** [comment]
I think one approach is to have two workflows which are similar to the two docker file approach:
- one is to pre-build Logseq and save the static artifacts with their version as the tag
- the other one is the executable/shared action workflow, which will download the required artifacts and then run the script

**Comment by Xuanwo** [comment]
Maybe it's better for us to just `docker run` directly.

**Comment by pengx17** [comment]
Ah, in some of my early attempts I did not manage to run the script outside of a docker build. 😂

**Comment by Xuanwo** [comment]
I didn't have a good idea on how to handle different versions of `logseq`/`logseq-publish`/`logsq-publish-action` so far.

Here are some quick ideas:

- Do we need to build a image for every logseq versions?
  - Or just `master`, `nightly` and releases versions?
- How to map the logseq versions to docker image tag?
  - One to one map looks good to you?

**Comment by pengx17** [comment]
I think we can keep it simple for now.
- The Docker build (`logseq-publish`) can be triggered manually without polling new upstream changes. This is because the published version (web only) is very stable compared to the app. So that there is no need to build it that often.
- We can also build Logseq version as well, e.g., by building a "0.5.8" tag (however this tag is not buildable because of the colors.js issue recently), when a new version is cut & published.

The user can choose which version of Logseq in his workflow but I guess it does not make much difference though. So I guess we can choose a version for them.

**Comment by Xuanwo** [comment]
The build result of test can be downloaded here: https://github.com/pengx17/logseq-publish/suites/4886440556/artifacts/139968121

I think this PR is almost ready for review!

PTAL @pengx17

**Comment by Xuanwo** [comment]
image

**Comment by Xuanwo** [comment]
I try to add `pages` for test, but the result is empty. Can you help debug it?

**Comment by pengx17** [comment]
> I try to add pages for test, but the result is empty. Can you help debug it?

Great! I will take a look today

**Comment by pengx17** [comment]
I think your approach works fine. I have tested to load a graph with a single `pages/contents.md` file. The exported `www` does not show contents in the page. Maybe you can try it again with your notes instead?

File: pages/contents.md [hunk 1]
```diff
@@ -0,0 +1 @@
+- Hello, World!
```
Comment (pengx17) [inline pages/contents.md]: Can you remove this file?

**Comment by Xuanwo** [comment]
It's quiet strange that the output index.html is empty.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+name: Test
+
+on: [push, pull_request]
```
Comment (pengx17) [inline .github/workflows/test.yml:3]: Can you mute this workflow?

File: action.yml [hunk 3]
```diff
@@ -17,11 +17,11 @@ branding:
...
-        container_id=$(docker create tmp)
-        docker cp $container_id:/home/logseq/graph-www ${{ inputs.dest }}
-        docker rm $container_id
-        rm -f Dockerfile
-      shell: bash
+    - uses: addnab/docker-run-action@v3
+      with:
+        image: "ghcr.io/pengx17/logseq-publish:${{ inputs.version }}"
+        options: |
+          -v ${{ github.workspace }}:/home/logseq/graph -v ${{ github.workspace }}/${{ inputs.dest }}:/home/logseq/graph-www
```
Comment (pengx17) [inline action.yml:27]: It seems if the `dest` folder is not created yet, it will be owned by root when mount to the container, see https://github.com/moby/moby/issues/13121#issuecomment-336186993.
Can you please add another step above to create the folder as current user if it does not exist?

**Comment by pengx17** [comment]
I have successfully published my graph using your refactored solution
https://github.com/pengx17/knowledge-garden/runs/4776296332?check_suite_focus=true

File: action.yml (hunk 3 above)
Comment (Xuanwo) [reply action.yml:27]: Added.

File: .github/workflows/test.yml (hunk 2 above)
Comment (Xuanwo) [reply .github/workflows/test.yml:3]: It's Ok, but why? I think it's better to make sure every commits tested via this workflow.

Comment (pengx17) [reply .github/workflows/test.yml:3]: Oh, you are right, let's keep it!
I suggest we can have a testing graph in this repo for testing instead of using the root. To support it we may need to add a new `input` parameter, like `context` or `graph_dir` etc.

File: action.yml (hunk 3 above)
Comment (pengx17) [reply action.yml:27]: Great, thank you!

File: .github/workflows/test.yml (hunk 2 above)
Comment (Xuanwo) [reply .github/workflows/test.yml:3]: That's great! I created an issue https://github.com/pengx17/logseq-publish/issues/6 to track it.

**Comment by Xuanwo** [comment]
All reviews have been addressed. @pengx17 PTAL.

**Review by pengx17 — APPROVED** [review]

**Comment by pengx17** [comment]
:shipit:

**Comment by Xuanwo** [comment]
Oops, closed?

**Comment by pengx17** [comment]
Sorry, miss clicked the wrong button ... Can you reopen a new PR?

**Comment by Xuanwo** [comment]
Thanks to git :laughing:
